WordPress/src/main/java/org/wordpress/android/ui/stockmedia/StockMediaPickerActivity.java

```diff
@@ -0,0 +1,606 @@
+            mSearchView.clearFocus();
+        }
+        ActivityUtils.hideKeyboard(this);
+        return true;
+    }
+
+    @Override
+    public boolean onQueryTextChange(String query) {
+        if (!StringUtils.equals(query, mSearchQuery)) {
+            submitSearch(query, true);
+        }
+        return true;
+    }
+
+    private void configureSearchView() {
+        mSearchView = findViewById(R.id.search_view);
+
+        // don't allow the SearchView to be closed
+        mSearchView.setOnCloseListener(new SearchView.OnCloseListener() {
+            @Override public boolean onClose() {
+                return true;
+            }
+        });
+
+        mSearchView.setOnQueryTextListener(this);
+    }
+
+    private void showEmptyView(boolean show) {
+        if (!isFinishing()) {
+            TextView txtEmpty = findViewById(R.id.text_empty);
+            txtEmpty.setVisibility(show ? View.VISIBLE : View.GONE);
+            if (show) {
+                boolean isEmpty = mSearchQuery == null || mSearchQuery.length() < MIN_SEARCH_QUERY_SIZE;
+                if (isEmpty) {
+                    String message = getString(R.string.stock_media_picker_initial_empty_text);
+                    String subMessage = getString(R.string.stock_media_picker_initial_empty_subtext);
+                    String link = "<a href='https://pexels.com/'>Pexels</a>";
+                    String html = message
+                                  + "<br /><br />"
+                                  + "<small>" + String.format(subMessage, link) + "</small>";
+                    txtEmpty.setMovementMethod(WPLinkMovementMethod.getInstance());
+                    txtEmpty.setText(Html.fromHtml(html));
+                } else {
+                    txtEmpty.setText(R.string.stock_media_picker_empty_results);
+                }
+            }
+        }
+    }
+
+    private void showProgress(boolean show) {
+        if (!isFinishing()) {
+            findViewById(R.id.progress).setVisibility(show ? View.VISIBLE : View.GONE);
+        }
+    }
+
+    private void submitSearch(@Nullable final String query, boolean delayed) {
+        mSearchQuery = query;
+
+        if (query == null || query.length() < MIN_SEARCH_QUERY_SIZE) {
+            mAdapter.clear();
+            showEmptyView(true);
+            return;
+        }
+
+        if (delayed) {
+            mHandler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    if (StringUtils.equals(query, mSearchQuery)) {
+                        submitSearch(query, false);
+                    }
+                }
+            }, 500);
+        } else {
+            fetchStockMedia(query, 1);
+        }
+    }
+
+    private void fetchStockMedia(@Nullable String searchQuery, int page) {
+        if (mIsFetching || !NetworkUtils.checkConnection(this)) {
+            return;
+        }
+
+        if (TextUtils.isEmpty(searchQuery)) {
```

Comment:
If the `searchQuery` is empty, we need to set `mIsFetching` to false. Since the search query changes and no new fetch starts, `onStockMediaListFetched` will never handle the callback and set it to `false`.

The issue can be reproduced (after fixing the issue in the previous comment) by starting a search for `dog` and deleting the query as soon as the fetch starts (might need to do it really fast depending on the connection) which will leave the progress bar spinning even though no fetch is going on.

It might be worth adding a comment for this call and explain why it's needed. Sorry for missing this one in the original review!

Reply:
I guess we also need to call the `showEmptyView(false)` in here.

I love that these type of things are not necessary with the observer pattern. ViewModel ❤️